ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/fullsync/FullSyncDownloadPipelineFactory.java

```diff
@@ -0,0 +1,116 @@
+import tech.pegasys.pantheon.ethereum.eth.sync.CheckpointHeaderFetcher;
+import tech.pegasys.pantheon.ethereum.eth.sync.CheckpointHeaderValidationStep;
+import tech.pegasys.pantheon.ethereum.eth.sync.CheckpointRangeSource;
+import tech.pegasys.pantheon.ethereum.eth.sync.DownloadBodiesStep;
+import tech.pegasys.pantheon.ethereum.eth.sync.DownloadHeadersStep;
+import tech.pegasys.pantheon.ethereum.eth.sync.DownloadPipelineFactory;
+import tech.pegasys.pantheon.ethereum.eth.sync.SynchronizerConfiguration;
+import tech.pegasys.pantheon.ethereum.eth.sync.ValidationPolicy;
+import tech.pegasys.pantheon.ethereum.eth.sync.state.SyncTarget;
+import tech.pegasys.pantheon.ethereum.mainnet.HeaderValidationMode;
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.metrics.MetricCategory;
+import tech.pegasys.pantheon.metrics.MetricsSystem;
+import tech.pegasys.pantheon.services.pipeline.Pipeline;
+import tech.pegasys.pantheon.services.pipeline.PipelineBuilder;
+
+import java.time.Duration;
+import java.util.Optional;
+
+public class FullSyncDownloadPipelineFactory<C> implements DownloadPipelineFactory {
+
+  private final SynchronizerConfiguration syncConfig;
+  private final ProtocolSchedule<C> protocolSchedule;
+  private final ProtocolContext<C> protocolContext;
+  private final EthContext ethContext;
+  private final MetricsSystem metricsSystem;
+  private final ValidationPolicy detachedValidationPolicy =
+      () -> HeaderValidationMode.DETACHED_ONLY;
+  private final BetterSyncTargetEvaluator betterSyncTargetEvaluator;
+
+  public FullSyncDownloadPipelineFactory(
+      final SynchronizerConfiguration syncConfig,
+      final ProtocolSchedule<C> protocolSchedule,
+      final ProtocolContext<C> protocolContext,
+      final EthContext ethContext,
+      final MetricsSystem metricsSystem) {
+    this.syncConfig = syncConfig;
+    this.protocolSchedule = protocolSchedule;
+    this.protocolContext = protocolContext;
+    this.ethContext = ethContext;
+    this.metricsSystem = metricsSystem;
+    betterSyncTargetEvaluator = new BetterSyncTargetEvaluator(syncConfig, ethContext.getEthPeers());
+  }
+
+  @Override
+  public Pipeline<?> createDownloadPipelineForSyncTarget(final SyncTarget target) {
+    final int downloaderParallelism = syncConfig.downloaderParallelism();
+    final int headerRequestSize = syncConfig.downloaderHeaderRequestSize();
+    final int singleHeaderBufferSize = headerRequestSize * downloaderParallelism;
+    final CheckpointRangeSource checkpointRangeSource =
+        new CheckpointRangeSource(
+            new CheckpointHeaderFetcher(
+                syncConfig, protocolSchedule, ethContext, Optional.empty(), metricsSystem),
+            this::shouldContinueDownloadingFromPeer,
+            ethContext.getScheduler(),
+            target.peer(),
+            target.commonAncestor(),
+            syncConfig.downloaderCheckpointTimeoutsPermitted(),
+            Duration.ofSeconds(5));
```

Comment:
Should this duration be stored as a constant somewhere?

Reply:
Done.